### DISCUSSION TITLE
Run secured archive services on a single host guide leads to archive not starting

### DISCUSSION THREAD

**Discussion by jmrpet** [dcm4che/dcm4chee-arc-light#3073] [Q&A] [answered]
**Describe the bug**
The https://github.com/dcm4che/dcm4chee-arc-light/wiki/Run-secured-archive-services-on-a-single-host
 guidance leads to the dcm4che/dcm4chee-arc-psql:5.23.1-secure docker container not starting.

I used the 'Use Docker Compose' approach.

**To Reproduce**
1. Go to instructions mentioned above
2. create the .env and .yml files as per instructions
3. run in powershell: docker-compose -p dcm4chee up -d
4. See error below. I think main issue is: java.lang.IllegalStateException: WFLYCTL0214: Could not get main file: dcm4chee-arc-oidc.xml. Specified files must be relative to the configuration dir: /opt/wildfly/standalone/configuration

`
Importing keystore /opt/wildfly/standalone/configuration/keystores/cacerts.p12 to /usr/local/openjdk-11/lib/security/cacerts...

Entry for alias mykey successfully imported.

Import command completed: 1 entries successfully imported, 0 entries failed or cancelled

Warning:

<mykey> uses a 1024-bit RSA key which is considered a security risk. This key size will be disabled in a future update.

Waiting for ldap:389 ...

done

Waiting for db:5432 ...

done

Starting Wildfly 22.0.0.Final

=========================================================================

JBoss Bootstrap Environment

JBOSS_HOME: /opt/wildfly

JAVA: /usr/local/openjdk-11/bin/java

JAVA_OPTS: -server -Xms64m -Xmx512m -XX:MetaspaceSize=96M -XX:MaxMetaspaceSize=256m -Djava.net.preferIPv4Stack=true -Djboss.modules.system.pkgs=org.jboss.byteman -Djava.awt.headless=true --add-exports=java.base/sun.nio.ch=ALL-UNNAMED --add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED --add-exports=jdk.unsupported/sun.reflect=ALL-UNNAMED

=========================================================================

java.lang.IllegalStateException: WFLYCTL0214: Could not get main file: dcm4chee-arc-oidc.xml. Specified files must be relative to the configuration dir: /opt/wildfly/standalone/configuration

at org.jboss.as.controller@14.0.0.Final//org.jboss.as.controller.persistence.ConfigurationFile.determineMainFile(ConfigurationFile.java:362)

at org.jboss.as.controller@14.0.0.Final//org.jboss.as.controller.persistence.ConfigurationFile.<init>(ConfigurationFile.java:200)

at org.jboss.as.server@14.0.0.Final//org.jboss.as.server.ServerEnvironment.<init>(ServerEnvironment.java:536)

at org.jboss.as.server@14.0.0.Final//org.jboss.as.server.Main.determineEnvironment(Main.java:388)

at org.jboss.as.server@14.0.0.Final//org.jboss.as.server.Main.main(Main.java:96)

at org.jboss.modules.Module.run(Module.java:353)

at org.jboss.modules.Module.run(Module.java:321)

at org.jboss.modules.Main.main(Main.java:617)`

**Expected behavior**
Expected all four containers to start after the 'docker-compose -p dcm4chee up -d' command but only three started.

**Desktop (please complete the following information):**
 - OS: Windows 10 as the host for docker

**Comment by gunterze** [comment, marked as answer]
Verify the content of the your host mount of `/opt/wildfly/standalone`, particular if `/opt/wildfly/standalone/configuration/dcm4chee-arc-oidc.xml` exists.
E.g.: with `docker-compose.yaml`
```yaml
  arc:
    image: dcm4che/dcm4chee-arc-psql:5.23.1-secure
    ...
    volumes:
      - /var/local/dcm4chee-arc/wildfly:/opt/wildfly/standalone
      - /var/local/dcm4chee-arc/storage:/storage
```

```console
$ ls -l /var/local/dcm4chee-arc/wildfly/configuration/dcm4chee-arc-oidc.xml
-rw-r--r-- 1 1023 1023 51556 Mär  8 15:24 /var/local/dcm4chee-arc/wildfly/configuration/dcm4chee-arc-oidc.xml
```

**Reply by jmrpet** [reply]
Thanks, @gunterze for the quick reply. 

You are right my host didn't have any files. Not sure did you notice I mentioned 'OS: Windows 10 as the host for docker' ?
Perhaps the instructions could mention that they are intended for a linux host and will not work on a windows host or do you think it could be made to work on windows host with the aproach I tried?

**My discoveries are below:**

I'm quite new to docker and I just followed the instructions as mentioned on: https://github.com/dcm4che/dcm4chee-arc-light/wiki/Run-secured-archive-services-on-a-single-host

So the docker-compose.yaml had the volumes listed as they are in the instructions. Not sure is this intended to work at all on a Windows host or do the instructions assume the user is running a linux host?

Reading up on volumes at https://docs.docker.com/compose/compose-file/compose-file-v3/#volume-configuration-reference
and seeing the format I think the yaml can't work on a windows host "The short syntax uses the generic [SOURCE:]TARGET[:MODE] format, where SOURCE can be either a host path or volume name" 

The row 75 in the yaml  - /var/local/dcm4chee-arc/wildfly:/opt/wildfly/standalone
would seem to assume my host has or could create a path /var/local/dcm4chee-arc/wildfly which I guess will not work on windows.

I guess a path relative to the compose file would work like this: 
- ./var/local/dcm4chee-arc/wildfly:/opt/wildfly/standalone

I tried changing the yaml so that all the /var ones were change to ./var and then all docker containers started.
Still, the "/etc/localtime:/etc/localtime:ro" rows I guess don't really work or at least it gave errors if I try the same approach.

I was able to login to keycloack and do the 'Register the Archive UI as OIDC client in Keycloak' step but accessing the https://<docker-host>:8443/dcm4chee-arc/ui2 wasn't possbile and returned an error. Perhaps there is something else which doesn't work on a windows host or keycloack needs the localtime and it's messed up due to the volumes part?

Note that the https://github.com/dcm4che/dcm4chee-arc-light/wiki/Run-minimum-set-of-archive-services-on-a-single-host instructions worked just fine even without any modifications in the yaml (sure the storage wasn't visible on the host...).

**Reply by jmrpet** [reply]
Did some more research. It seems I hadn't installed a distro under the wsl2. This was step 6 in https://docs.microsoft.com/en-us/windows/wsl/install-win10. After installing wsl2 ubuntu 20.04 and enabling it in docker it seems that the docker containers started just fine with the original yaml

Found this article helpful in finding where the stuff gets mounted to
https://stackoverflow.com/questions/61083772/where-are-docker-volumes-located-when-running-wsl-using-docker-desktop

In my case mounts went to \\wsl$\docker-desktop\tmp\docker-desktop-root\containers\services\docker\tmp\upper\var\local\dcm4chee-arc\wildfly

**Reply by jmrpet** [reply]
I'm starting to realize perhaps this should have been recorded as a question in the google groups side.
Perhaps still something could be mentioned in the wiki to clarify can this be run cross-platform in docker and perhaps some notes on how?

It seems after step 6 when opening the url https://:8443/dcm4chee-arc/ui2 it gives an error and 

wildfly server.log shows the below. 

2021-03-08 23:26:11,498 WARN  [org.keycloak.adapters.KeycloakDeployment] (default task-1) Failed to load URLs from https://dell-laptop:8843/auth/realms/dcm4che/.well-known/openid-configuration: java.net.UnknownHostException: dell-laptop: No address associated with hostname

Perhaps this is some networking issue in the wsl2 Ubuntu

Launching ubuntu directly from start menu and doing a ping does give the below result however
ping dell-laptop
PING Dell-laptop.localdomain (127.0.1.1) 56(84) bytes of data.

**Reply by gunterze** [reply]
Seems `dell-laptop` is not registered in your DNS, so it can't be resolved  from inside the container. You may workaround by adding a mapping of `dell-laptop` to the IP of the docker host - **not** 127.0.1.1! - in the host file of the container by https://github.com/compose-spec/compose-spec/blob/master/spec.md#extra_hosts .
